pkg/reconciler/git/client.go

```diff
@@ -26,6 +26,81 @@ func (d *Client) Worktree() (*git.Worktree, error) {
 	return d.repo.Worktree()
 }
 
-func (d *Client) ResolveRevision(rev gitp.Revision) (*gitp.Hash, error) {
+func (d *Client) ResolveRevisionOrBranchHead(rev gitp.Revision) (*gitp.Hash, error) {
+	// no version provided. We will set rev to the default branch name.
+	if rev.String() == "" {
+		branch, err := d.DefaultBranch()
+		if err != nil {
+			return nil, err
+		}
+		rev = gitp.Revision(branch.Name().Short())
+	}
+	// is rev a branch name? We can't use repo.Branch() because that checks local branches only :/
+	// So, we try to resolve rev as a remote branch reference. If that works, we
+	// can just use it 's head!
+	branchRev, err := d.repo.ResolveRevision(
+		gitp.Revision(
+			fmt.Sprintf("refs/remotes/origin/%s", rev.String()),
+		),
+	)
+	if err != nil && err != gitp.ErrReferenceNotFound {
+		return nil, err
+	}
+	if branchRev != nil { // this is a branch
+		return d.repo.ResolveRevision(gitp.Revision(branchRev.String()))
+	}
 	return d.repo.ResolveRevision(rev)
 }
+
+func (d *Client) Fetch(path string, o *git.FetchOptions) error {
+	var err error
+	if d.repo == nil {
```

Comment:
A couple of questions here:
- Why fetch method has a `path` argument?
- Why this methods opens the repository from the `path`?
- What the user of this method should expect if he will provide both: `a repository` and `a path`?

Reply:
1- The path is used in case the client object is not initialized with a repo already.
2- Because `fetch` requires a cloned local copy of the repo. We need to "open" the local repo cache. 
3- Repository will take precedence. Since the the repository is not passed, it's only initialized if this client instance is used to clone the repo from upstream. In which case we don't need to "open" it again.

Reply:
The git library need in general a "refurbishment" as it's already bad designed. The `Client` struct is in general unnecessary I think.

Maybe better to fail always when `d.repo == nil` with an unexpected state error.

And initialising the repo has to happen upfront by e.g. calling `NewRepositoryFromPath(path string) (*git.Repository, error)` and return the *git.Repository as result.

Reply:
=> follow up action to clean up the GIT library in general is https://github.com/kyma-incubator/reconciler/issues/432